### PR TITLE
[WIP] Graph builder for DALI pipeline test

### PR DESCRIPTION
**Pull request by mzient**
Concept of a graph builder for DALI operator/pipeline test.
Usage:
```
  OpDAG dag;
  auto &n1 = dag.add("MyOp1")(dag.in[0]);
  auto &n2 = dag.add("MyOp2")(dag.in[1]);
  auto &n3 = dag.add("MyOp3")(n1[0], n2[0]);
  dag.out(n3[0]);
```
Simple use cases:
```
OpDAG dag = SingleOpGraph("MyOp", Node::CPU);  // single operator, 1 input, 1 output
OpDAG dag = SingleOpGraph("MyOp", Node::CPU, 2, 3);  // single operator, 2 inputs, 3 outputs
```

Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

### CODE REVIEW

**Comment by JanuszL** [comment]
What about providing arguments for graph operators?
Another question is we want to rework DALI to use that approach anywhere else than tests?

**Comment by mzient** [comment]
@JanuszL What I wanted to achieve here was separating the pipeline definition from arguments, so we can run pipelines with different sets of arguments using e.g. `TEST_P`. This graph doesn't build any pipeline - it will be merged with scalar argument set to produce OpScpecs that will define the pipeline. I also want to make specifying arguments and `ArgumentInput`s uniform.

File: dali/test/op_graph.h [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  NodeOutput gpu() const {
+    auto ret = *this;
+    ret.explicit_GPU = true;
+    return ret;
+  }
+};
+
+struct Node {
+  Node() = default;
+  Node(const Node &) = delete;
```
Comment (szalpal) [inline dali/test/op_graph.h]: I just wanted to point out, that here's a circular dependency (`Node <-> Edge`); and I heard it's a bad practice

File: dali/test/op_graph.h [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    return true;
+  }
+
+  std::unordered_map<std::string, Node> nodes;
+  Node in{"", "__input", Node::AnyBackend};
+  Node out{"", "__output", Node::AnyBackend};
+};
+
+
+OpDAG SingleOpGraph(const std::string &op_name, int backend_mask = Node::AnyBackend, int num_inputs = 1, int num_outputs = 1) {
```
Comment (szalpal) [inline dali/test/op_graph.h]: Isn't `num_inputs` and `num_outputs` reundant? I mean, we can figure it out on `OpSpec` creating level and it seems, we don't need it earlier?

Comment (mzient) [reply dali/test/op_graph.h]: Maybe we can make 0 a special value (the number of inputs/outputs will be inferred from arguments supplied to Run).

File: dali/test/op_graph.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+  template <typename... Inputs>
+  Node &operator()(Inputs&&... inputs) {
+    int junk[] = { (add_in(inputs), 0)... };
+    (void)junk;
+    return *this;
+  }
+
+  void add_in(Edge e) {
+    in.push_back(e);
```
Comment (JanuszL) [inline dali/test/op_graph.h]: Don't we need to move this `in.push_back(e);` to the end after we modify e, not before?

Comment (JanuszL) [reply dali/test/op_graph.h]: Oh, I see, we modify only things pointed by node.

File: dali/test/op_graph.h [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  std::string op_name;
+  std::string node_name;
+  int backend_mask = 0;
+  int num_outputs = 0;
+  std::vector<Edge> in;
+  std::vector<Edge> out;
+};
+
+struct OpDAG {
+  Node &add(std::string op_name, std::string node_name, int backend_mask = Node::AnyBackend) {
```
Comment (szalpal) [inline dali/test/op_graph.h]: According to Google Style (which we use), [this should be a](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) `class`. Also, [the name should be](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) `OpDag` (I'd prefer `OperatorDag` though)

File: dali/test/op_graph.h (hunk 1 above)
Comment (mzient) [reply dali/test/op_graph.h]: There isn't. Edge doesn't depend on definition of Node. Even Edge's member function bodies don't.

**Comment by mzient** [comment]
@klecki I fixed it and will dismiss your previous review as outdated.

File: dali/test/op_graph.h (hunk 4 above)
Comment (szalpal) [reply dali/test/op_graph.h]: > < rant >I really do hate the practice of decapitalizing acronyms. < /rant >

I don't wanna be this kind of guy that pick holes or anything ;) It's just in my opinion, there's a purpose, why we agreed to use the Google Style - consistency. I'm not saying, that it's unchangeable, it's just the whole point of consistency is that everybody writes a code in the same way. Anywho, I'm totally open for any discussion for code guidelines :)

Comment (mzient) [reply dali/test/op_graph.h]: Making it a `class` would call for the ridiculous underscore-suffixed members - and possibly making them private.
I want to have public `in` and `out` in the DAG object. However, I can use a NOLINT for that...
Making `in` and `out` functions doesn't play well with the rest of the syntax when we use both `operator[]` for extracting outputs  and `operator()` for assigning inputs. Actually, `out` could be a function, but `in` cannot - and I'd rather keep them consistent.

**Pull request closed without merging**